### PR TITLE
Add generics for Remix type enhancements

### PR DESCRIPTION
**Pull request by brophdawg11**
Adding TS generics to support https://github.com/remix-run/remix/pull/7319

### CODE REVIEW

File: packages/router/router.ts [hunk 1]
```diff
@@ -394,20 +395,12 @@ export interface RouterSubscriber {
   (state: RouterState): void;
 }
 
-interface UseMatchesMatch {
-  id: string;
-  pathname: string;
-  params: AgnosticRouteMatch["params"];
-  data: unknown;
-  handle: unknown;
-}
```
Comment (brophdawg11) [inline packages/router/router.ts:403]: This is now exported from `utils.ts` as `UIMatch`

File: packages/router/history.ts [hunk 2]
```diff
@@ -49,15 +49,18 @@ export interface Path {
...
 
+// TODO: (v7) Change the Location generic default from `any` to `unknown` and
+// remove Remix `useLocation` wrapper.
+
 /**
  * An entry in a history stack. A location contains information about the
  * URL path, as well as possibly some arbitrary state and a key.
  */
-export interface Location extends Path {
+export interface Location<S = any> extends Path {
```
Comment (MichaelDeBoey) [inline packages/router/history.ts:59]:
```suggestion
export interface Location<State = any> extends Path {
```

File: packages/router/utils.ts [hunk 3]
```diff
@@ -137,21 +137,24 @@ export type Submission =
...
 }
 
+// TODO: (v7) Change the defaults from any to unknown in and remove Remix wrappers:
+//   ActionFunction, ActionFunctionArgs, LoaderFunction, LoaderFunctionArgs
+
 /**
  * Arguments passed to loader functions
  */
-export interface LoaderFunctionArgs extends DataFunctionArgs {}
+export interface LoaderFunctionArgs<C = any> extends DataFunctionArgs<C> {}
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:152]:
```suggestion
export interface LoaderFunctionArgs<Context = any> extends DataFunctionArgs<Context> {}
```

File: packages/router/utils.ts [hunk 4]
```diff
@@ -137,21 +137,24 @@ export type Submission =
...
 
 /**
  * Arguments passed to action functions
  */
-export interface ActionFunctionArgs extends DataFunctionArgs {}
+export interface ActionFunctionArgs<C = any> extends DataFunctionArgs<C> {}
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:157]:
```suggestion
export interface ActionFunctionArgs<Context = any> extends DataFunctionArgs<Context> {}
```

File: packages/router/utils.ts [hunk 5]
```diff
@@ -163,15 +166,15 @@ type DataFunctionValue = Response | NonNullable<unknown> | null;
 /**
  * Route loader function signature
  */
-export interface LoaderFunction {
-  (args: LoaderFunctionArgs): Promise<DataFunctionValue> | DataFunctionValue;
+export interface LoaderFunction<C = any> {
+  (args: LoaderFunctionArgs<C>): Promise<DataFunctionValue> | DataFunctionValue;
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:170]:
```suggestion
export interface LoaderFunction<Context = any> {
  (args: LoaderFunctionArgs<Context>): Promise<DataFunctionValue> | DataFunctionValue;
```

File: packages/router/utils.ts [hunk 6]
```diff
@@ -163,15 +166,15 @@ type DataFunctionValue = Response | NonNullable<unknown> | null;
...
 }
 
 /**
  * Route action function signature
  */
-export interface ActionFunction {
-  (args: ActionFunctionArgs): Promise<DataFunctionValue> | DataFunctionValue;
+export interface ActionFunction<C = any> {
+  (args: ActionFunctionArgs<C>): Promise<DataFunctionValue> | DataFunctionValue;
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:177]:
```suggestion
export interface ActionFunction<Context = any> {
  (args: ActionFunctionArgs<Context>): Promise<DataFunctionValue> | DataFunctionValue;
```

File: packages/router/utils.ts [hunk 7]
```diff
@@ -490,6 +493,28 @@ export function matchRoutes<
   return matches;
 }
 
+export interface UIMatch<D = unknown, H = unknown> {
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:496]:
```suggestion
export interface UIMatch<Data = unknown, Handle = unknown> {
```

File: packages/router/utils.ts [hunk 8]
```diff
@@ -490,6 +493,28 @@ export function matchRoutes<
...
+  id: string;
+  pathname: string;
+  params: AgnosticRouteMatch["params"];
+  data: D;
+  handle: H;
```
Comment (MichaelDeBoey) [inline packages/router/utils.ts:501]:
```suggestion
  data: Data;
  handle: Handle;
```

**Review by MichaelDeBoey — COMMENTED** [review]
In the light of named tuple elements, I've seen something passing by to do the same for generic elements as well, so I think it would be better to have long names instead

https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#named-and-anonymous-tuple-elements

**Pull request merged**